consensus/core/src/dag_state.rs

```diff
@@ -257,6 +257,48 @@ impl DagState {
         Ok(blocks)
     }
 
+    /// Returns the last block proposed per authority with round <= `before_round`. The method will look
+    /// into cache first, otherwise will fall back in store. The method guarantees to return a result for
+    /// each authority, even if that's genesis block.
+    pub(crate) fn get_last_block_per_authority(
+        &self,
+        before_round: Round,
```

Comment:
When `before_round` is within 1~3 rounds of highest accepted round, highest blocks below `before_round` must be in `recent_blocks` and this function should never need to hit disk. Any authority without a block in recent blocks should not have produced blocks after genesis. But providing a generic implementation like below could be less error prone.

Reply:
Absolutely agree. The primary reason why I've added the disk fallback is for API completeness - so someone can use this method with the confidence that it will always yield the correct result irrespective of the provided `before_round`. Now, as you said, in proposal we should never hit disk.